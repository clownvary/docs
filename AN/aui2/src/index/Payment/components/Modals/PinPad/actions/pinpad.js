import { deepMerge } from 'shared/utils/func';
import { addError } from 'shared/actions/Error';

import getUrlsForPinpad from './pinpadHelper';
import { showModal, updateModalTitle } from './pinpadModal';
import { showFailActionBtns } from './pinpadFail';
import { titles, actionTypes } from '../consts';
import { getCardTypeSystemIdByName, getCardTypeIdByCardTypeName } from '../../../../utils/creditCardHelper';

const URL = getUrlsForPinpad();
export const GO_TO_NEXT_PAGE = 'GO_TO_NEXT_PAGE';
export const INIT_DATA_OF_PINPAD = 'INIT_DATA_OF_PINPAD';

// Only use in new device logic
export const GET_APD_INFORMATION = 'GET_APD_INFORMATION';
export const GET_APD_INFORMATION_SUCCESS = 'GET_APD_INFORMATION_SUCCESS';
export const GET_APD_INFORMATION_FAILURE = 'GET_APD_INFORMATION_FAILURE';

// Only use in new device logic
export const GET_APD_SERVER_STATUS = 'GET_APD_SERVER_STATUS';
export const GET_APD_SERVER_STATUS_SUCCESS = 'GET_APD_SERVER_STATUS_SUCCESS';
export const GET_APD_SERVER_STATUS_FAILURE = 'GET_APD_SERVER_STATUS_FAILURE';

// Only use in new device logic
export const PROCESS_APD_RESPONSE = 'PROCESS_APD_RESPONSE';
export const PROCESS_APD_RESPONSE_SUCCESS = 'PROCESS_APD_RESPONSE_SUCCESS';
export const PROCESS_APD_RESPONSE_FAILURE = 'PROCESS_APD_RESPONSE_FAILURE';

export const GET_APD_PAYMENT_INFO = 'GET_APD_PAYMENT_INFO';
export const GET_APD_PAYMENT_INFO_SUCCESS = 'GET_APD_PAYMENT_INFO_SUCCESS';
export const GET_APD_PAYMENT_INFO_FAILURE = 'GET_APD_PAYMENT_INFO_FAILURE';

export const COMMUNICATE_WITH_DEVICE = 'COMMUNICATE_WITH_DEVICE';
export const UPDATE_PROMPT_MESSAGE = 'UPDATE_PROMPT_MESSAGE';
export const UPDATE_ERROR_MESSAGE = 'UPDATE_ERROR_MESSAGE';
export const UPDATE_SUCCESS_MESSAGE = 'UPDATE_SUCCESS_MESSAGE';
export const ACCOUNT_HOLDER_INFO = 'ACCOUNT_HOLDER_INFO';
export const CLEAR_PINPAD_PAYS = 'CLEAR_PINPAD_PAYS';
export const UPDATE_CURRENT_PAY_INDEX = 'UPDATE_CURRENT_PAY_INDEX';
export const UPDATE_SUCCESS_PAYMENT = 'UPDATE_SUCCESS_PAYMENT';

export const PINPAD_SET_IS_NEW_CARD_FLAG = 'PINPAD_SET_IS_NEW_CARD_FLAG';

let nextPageFunc = () => ({
  type: GO_TO_NEXT_PAGE,
  payload: {
    value: false
  }
});

let _resolve = () => {};
let _reject = () => {};

const setIsNewCard = bFlag => ({
  type: PINPAD_SET_IS_NEW_CARD_FLAG,
  payload: { bFlag }
});

export function initDataOfPinpad(data) {
  return {
    type: INIT_DATA_OF_PINPAD,
    payload: { data }
  };
}

// Only use in new device logic
export function getAPDInformation() {
  return {
    types: [GET_APD_INFORMATION, GET_APD_INFORMATION_SUCCESS, GET_APD_INFORMATION_FAILURE],
    promise: API => API.get(URL.getAPDInfo, {
      body: { is_test: true }
    })
  };
}

const getCreditCardexpireInfo = walletId => ({
  types: ['', '', ''],
  promise: API => API.get(URL.ccexpireinfo, {
    body: { wallet_id: walletId }
  })
});

// Only use in new device logic
function _getAPDServerStatus({ batchID, receiptID }) {
  return {
    types: [GET_APD_SERVER_STATUS, GET_APD_SERVER_STATUS_SUCCESS, GET_APD_SERVER_STATUS_FAILURE],
    promise: API => API.get(URL.getAPDServerStatus, {
      body: {
        batch_id: batchID,
        receipt_id: receiptID,
        is_wallet_save: false
      }
    }),
    meta: {
      ignoreLoadingbar: true
    }
  };
}

function _receiptCommitError(value) {
  // Set error on the payment page. Use random error code > 1000
  // so error will be detected as business error
  // and will display at page top rather than as system error which will display in popup
  return (dispatch) => {
    dispatch(addError({
      payload: {
        code: '3000',
        message: value
      }
    }));
  };
}

// Only use in new device logic
export function getAPDServerStatus(params, successCallback, errorCallback) {
  return dispatch => dispatch(_getAPDServerStatus(params))
    .then(({ payload: { body: { apd_server_status } } }) => {
      successCallback && successCallback(apd_server_status);
    }, ({ payload: { headers: { response_code: responseCode } } }) => {
      // istanbul ignore else
      if (responseCode === '0002') {
        errorCallback && errorCallback();
      }
    });
}

// Only use in new device logic
function _processAPDResponse(params) {
  return {
    types: [PROCESS_APD_RESPONSE, PROCESS_APD_RESPONSE_SUCCESS, PROCESS_APD_RESPONSE_FAILURE],
    promise: API => API.post(URL.processAPDResponse, {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(params)
    }),
    meta: {
      ignoreLoadingbar: true
    }
  };
}

// Only use in new device logic
export function processAPDResponse(params, successCallback, failCallback) {
  return dispatch => dispatch(_processAPDResponse(params))
    .then(({ payload: { body: { apd_process_response } } }) => {
      successCallback && successCallback(apd_process_response);
    }, () => {
      failCallback && failCallback();
    });
}

export function getAPDPaymentInfo({ batchID, receiptID }) {
  return {
    types: [GET_APD_PAYMENT_INFO, GET_APD_PAYMENT_INFO_SUCCESS, GET_APD_PAYMENT_INFO_FAILURE],
    promise: API => API.get(URL.getAPDPaymentInfo, {
      body: {
        batch_id: batchID,
        receipt_id: receiptID
      }
    })
  };
}

export function setAccountHolder({ payload: { body: { account_holder } } }) {
  return {
    type: ACCOUNT_HOLDER_INFO,
    payload: {
      value: account_holder
    }
  };
}

export function communicatingWithDevice(flag) {
  return {
    type: COMMUNICATE_WITH_DEVICE,
    payload: {
      value: flag
    }
  };
}

export function gotoNextPage(receiptHeaderId) {
  return dispatch => dispatch(nextPageFunc(receiptHeaderId));
}

// Always keep the credit card above the debit card if have both.
function sortPays(initPinpadData) {
  const pinpadData = deepMerge({}, initPinpadData);
  const { pays, debitCardId } = pinpadData;
  const finalPays = deepMerge([], pays);

  if ((pays.length > 1) && (pays[0].paymentTypeId === debitCardId)) {
    finalPays[0] = pays[1];
    finalPays[1] = pays[0];
  }

  pinpadData.pays = finalPays;

  return pinpadData;
}

export function swipeCardInPinpadModal(dataObj, actions) {
  const { fetchAccountHolderAction, gotoNextPage: next } = actions;
  nextPageFunc = next;
  const data = sortPays(deepMerge({}, dataObj));

  return dispatch => Promise.all([
    dispatch(initDataOfPinpad(data)),
    __STATIC__ ? dispatch(getAPDInformation()) : null, // Only use in new device logic
    dispatch(getAPDPaymentInfo(data)),
    dispatch(fetchAccountHolderAction())
  ]).then((args) => {
    const accrountHolderResult = args[3];
    dispatch(setAccountHolder(accrountHolderResult));
    dispatch(setIsNewCard(false));
    dispatch(updateModalTitle(titles.TRANSACTION_TITLE));
    dispatch(showModal(true));
    dispatch(communicatingWithDevice(true));
  });
}

export const processResolveAction = ({ walletId, ccMasked, ccCardTypeName }) =>
  (dispatch, getState) => dispatch(getCreditCardexpireInfo(walletId))
  .then(({
    payload: {
      body: { cc_exp: expirationDate }
    }
  }) => {
    const { cardTypeList } = getState().paymentModals.newCreditCard.toJS();
    const cardTypeSystemId = getCardTypeSystemIdByName(ccCardTypeName, cardTypeList.data);
    const cardTypeId = getCardTypeIdByCardTypeName(ccCardTypeName, cardTypeList.data);

    dispatch(showModal(false));

    _resolve({
      walletId,
      ccNumberValue: ccMasked,
      cardTypeId,
      cardTypeSystemId,
      expirationDate
    });
  });

export function getCardInfoByPinPadPromiseAction(dataObj, actions) {
  const { fetchAccountHolderAction, gotoNextPage: next } = actions;
  nextPageFunc = next;
  const data = sortPays(deepMerge({}, dataObj));

  return dispatch => Promise.all(
    [
      dispatch(initDataOfPinpad(data)),
      dispatch(fetchAccountHolderAction())
    ])
    .then(args => new Promise(
      (resolve, reject) => {
        _resolve = resolve;
        _reject = reject;

        const accrountHolderResult = args[1];
        dispatch(setAccountHolder(accrountHolderResult));
        dispatch(setIsNewCard(true));
        dispatch(updateModalTitle(titles.NEW_CARD_TITLE));
        dispatch(showModal(true));
        dispatch(communicatingWithDevice(true));
      }));
}

export const processRejectAction = () =>
  () => _reject();

export function updateCurrentPayIndex(index) {
  return {
    type: UPDATE_CURRENT_PAY_INDEX,
    payload: {
      value: index
    }
  };
}

export function updatePromptMessage(message, actionType = actionTypes.HIDE_RETRY) {
  return {
    type: UPDATE_PROMPT_MESSAGE,
    payload: {
      message,
      actionType
    }
  };
}

function updateErrMsg({ message, isServerErr, hasError, actionType = actionTypes.SHOW_RETRY }) {
  return {
    type: UPDATE_ERROR_MESSAGE,
    payload: {
      message,
      isServerErr,
      actionType,
      hasError // should set hasError to false when retry
    }
  };
}

export function updateErrorMessage(obj) {
  return (dispatch, getState) => Promise.all([
    dispatch(updateErrMsg(obj))
  ]).then(() => {
    const {
      pays,
      errorMessage
    } = getState().paymentModals.pinpad.payment.toJS();

    const paysNotSwipe = pays.filter(item => !item.init);

    // istanbul ignore else
    if (paysNotSwipe.length) {
      if (errorMessage) {
        dispatch(updateErrMsg({
          message: '',
          isServerErr: true
        }));
      }

      dispatch(showFailActionBtns(false));
      dispatch(communicatingWithDevice(true));
      dispatch(updateCurrentPayIndex(pays.indexOf(paysNotSwipe[0])));
    }
  });
}

export function notifySuccessPayment({ index }) {
  return {
    type: UPDATE_SUCCESS_PAYMENT,
    payload: {
      index
    }
  };
}

function updateSuccessMsg(message) {
  return {
    type: UPDATE_SUCCESS_MESSAGE,
    payload: {
      message
    }
  };
}


export function updateSuccessMessage({ message, finishReceiptErrorMessage }) {
  return (dispatch, getState) => {
    const {
      currentPayIndex,
      pays: oldPays
    } = getState().paymentModals.pinpad.payment.toJS();
    const curPay = oldPays[currentPayIndex];

    return Promise.all([
      dispatch(updateSuccessMsg(message)),
      dispatch(notifySuccessPayment(curPay))
    ]).then(() => {
      const {
        pays,
        receiptHeaderId
      } = getState().paymentModals.pinpad.payment.toJS();

      const isAllPaysSuccess = pays.every(item => item.swipeSuccess);

      if (isAllPaysSuccess) {
        if (finishReceiptErrorMessage && finishReceiptErrorMessage.length) {
          // Close the pinpad modal window now and return to the payment page
          dispatch(showModal(false));
          dispatch(_receiptCommitError(finishReceiptErrorMessage));
        } else {
          // Continue on to the permit confirmation page
          dispatch(gotoNextPage(receiptHeaderId));
        }
      }

      const paysNotSwipe = pays.filter(item => !item.init);

      if (paysNotSwipe.length) {
        dispatch(communicatingWithDevice(true));
        dispatch(updateCurrentPayIndex(pays.indexOf(paysNotSwipe[0])));
      } else {
        dispatch(showFailActionBtns(true));
      }

      return false;
    });
  };
}


export function clearPinpadPays() {
  return {
    type: CLEAR_PINPAD_PAYS
  };
}
