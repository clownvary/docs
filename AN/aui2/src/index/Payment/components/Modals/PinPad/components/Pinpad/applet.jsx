import React from 'react';
import UIComponent from 'shared/components/UIComponent';
import './index.less';

const navigatorName = window.navigator.appName.toLowerCase();
const isIE = navigatorName.indexOf('internet explorer') !== -1;
const isNetScape = navigatorName.indexOf('netscape') !== -1;
/* eslint-disable */
export default class Applet extends UIComponent {
  render() {
    return (
      <div
        className="pinpad-transaction-fail"
        dangerouslySetInnerHTML={{ __html: this.renderApplet() }}
      />
    );
  }

  renderApplet() {
    const { applet_name, applets } = this.props.appletInfo;

    if (isIE || isNetScape) {
      if (isNetScape) {
        return `<applet
          ${applets}
          archive=${applet_name}.jar
          code=${applet_name}.class
          id=APDInterfaceApplet
          name=APDInterfaceApplet
          height=0
          width=0
          scriptable=true
          mayscript=mayscript
          pluginspage=http://java.com/en/download/index.jsp>
          ${this.renderParams()}
        </applet>`;
      }
      return `<object
        codebase={code_base}
        type=application/x-java-applet;version=1.5
        archive=${applet_name}.jar
        code=${applet_name}.class
        id=APDInterfaceApplet
        name=APDInterfaceApplet
        height=0
        width=0>
          <param name=mayscript value=true/>
          ${this.renderParams()}
        </object>`;
    }
    return '';
  }

  renderParams() {
    const {
      device_type_id,
      terminal_location_address,
      device_id,
      terminal_location_city,
      terminal_location_state,
      terminal_location_zip,
      card_acceptor_name,
      merchant_descriptor,
      merchant_user,
      merchant_password,
      ams_url,
      ams_port,
      ams_timeout,
      terminal_number,
      usb_port_name,
      pinpad_data_entry_timeout,
      pinpad_swipe_card_timeout,
      pinpad_pin_entry_timeout,
      write_to_log
    } = this.props.appletInfo;

    const ccScanWithoutEmv = device_type_id === 1 || device_type_id === 2;
    return `${ccScanWithoutEmv ? '<param name="separate_jvm" value="true"><param name="cache_option" value="No"><param name="classloader_cache" value="true">' : ''}
      <param name="dll_ver" value="1.0.0">
      <param name="dll_auto_update" value="Yes">
      <param name="AMSURL" value="${ams_url}">
      <param name="AMSPort" value="${ams_port}">
      <param name="AMSTimeout" value="${ams_timeout}">
      <param name="CardAcceptorName" value="${card_acceptor_name}">
      <param name="TerminalLocationAddress" value="${terminal_location_address}">
      <param name="TerminalLocationCity" value="${terminal_location_city}">
      <param name="TerminalLocationState" value="${terminal_location_state}">
      <param name="TerminalLocationZIP" value="${terminal_location_zip}">
      <param name="MerchantDescriptor" value="${merchant_descriptor}">
      <param name="MerchantUser" value="${merchant_user}">
      <param name="MerchantPassword" value="${merchant_password}">
      <param name="TerminalNumber" value="${terminal_number}">
      <param name="DeviceId" value="${device_id}">
      <param name="USBPortName" value="${usb_port_name}">
      <param name="PinpadDataEntryTimeout" value="${pinpad_data_entry_timeout}">
      <param name="PinpadPINEntryTimeout" value="${pinpad_pin_entry_timeout}">
      <param name="PinpadSwipeCardTimeout" value="${pinpad_swipe_card_timeout}">
      <param name="WriteToLog" value="${write_to_log}">`;
  }
}
/* eslint-enable */


/*
  the relationship between pinpad device type with applet name:
  {
    "1": "APDInterfaceApplet", //paymentech
    "2": "APDMonerisInterfaceApplet",
    "3": "APDEMVInterfaceApplet", //paymentech EMV
    "4": "APDEMVElavonInterfaceApplet"
  }
*/
