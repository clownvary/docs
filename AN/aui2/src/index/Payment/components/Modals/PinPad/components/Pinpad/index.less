@import '~shared/style/variables/index.less';
@imagePath: '~shared/images/iPP320.jpg';
@module-name: payment-pinpad;

.@{module-name} {
  p {
    padding: 0;
    margin: 0;
  }

  &-header {
    padding-bottom: 10px;
    overflow: hidden;
    border-bottom: 1px solid #d6d7da;

    span {
      &:first-child {
        float: left;
        font-weight: @semibold;
      }

      &:nth-child(2) {
        float: right;
        font-weight: 700;
      }
    }
  }

  &-body {
    padding-top: 30px;

    &.aaui-flexbox {
      align-items: flex-start;
      flex-wrap: nowrap;
    }

    &-img {
      width: 190px;
      height: 190px;
      font-size: 0;
      background-image: url(@imagePath);
      background-position: -20px 0;
      background-repeat: no-repeat;
      background-size: 100%;
    }

    &-right {
      height: 190px;
    }

    &-prompts {
      color: #748288;
    }
  }

  &-error-wrapper {
    &.aaui-flexbox {
      align-items: flex-end;
      flex-direction: column;
      flex-wrap: nowrap;
      justify-content: space-between;
      height: 100%;
    }

    .@{module-name}-retry {
      cursor: pointer;
    }

    .@{module-name}-error-alert {
      display: flex;
      width: 398px;

      .icon-close-circle,
      .icon-times-circle {
        padding-left: 10px;
        margin-top: 0;
        font-size: 16px;
        color: #db3811;
      }

      .error-content {
        display: inline-block;
        padding-right: 10px;
        padding-left: 10px;
        margin-top: -3px;
        vertical-align: top;
      }

      ul {
        padding: 0;
        margin: 0;
        list-style-type: none;
      }

      .@{module-name}-error-message {
        max-height: 65px;
        overflow: auto;
        font-size: 14px;
        line-height: 1.43;
      }
    }

    .alert-error {
      color: #db3811;
    }
  }

  &-error-action {
    line-height: 1.43;
    color: #748288;

    .payment-pinpad-retry {
      color: @color-blue;
    }

    .payment-pinpad-retry-separation {
      margin: 0 5px;
    }
  }
}
