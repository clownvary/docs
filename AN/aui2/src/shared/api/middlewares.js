import PromiseMiddleware from './promiseMiddleware';
import RouteMiddleware from './routeMiddleware';
import ThunkMiddleware from './thunkMiddleware';
import API from './API';

export default [
  new ThunkMiddleware(),
  new RouteMiddleware(),
  new PromiseMiddleware(new API())
];
