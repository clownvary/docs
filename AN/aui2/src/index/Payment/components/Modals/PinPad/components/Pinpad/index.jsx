import Pinpad from './pinpad';
import SinglePay from './singlePay';
import MutiplePays from './mutiplePays';

import './index.less';

export default {
  Pinpad,
  SinglePay,
  MutiplePays
};
