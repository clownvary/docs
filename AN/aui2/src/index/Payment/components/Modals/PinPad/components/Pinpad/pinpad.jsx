import React from 'react';
import { connect } from 'react-redux';
import Alert from 'react-base-ui/lib/components/Alert';
import UIComponent from 'shared/components/UIComponent';
import { WSIHelper } from 'shared/components/WSIHelper';
import { updateErrorMessage } from '../../actions/pinpad';
import { showFailActionBtns } from '../../actions/pinpadFail';
import { updateModalTitle } from '../../actions/pinpadModal';
import { titles, actionTypes } from '../../consts';

export class Pinpad extends UIComponent {

  render() {
    const { pay, pinpad, children } = this.props;
    let { onClick } = this.props;
    const { promptMessage, debitCardId, isNewCard } = pinpad.payment.toJS();
    const {
      payName,
      errorMessage,
      actionType,
      hasError,
      amount,
      headerHtml,
      paymentTypeId,
      className
    } = pay;
    onClick = onClick || (() => {});

    return (
      <div className={`payment-pinpad ${className || ''}`} onClick={onClick}>
        {
          !isNewCard &&
          <div className="payment-pinpad-header">
            <span>{payName}</span>
            <span>$ {amount}</span>
            <div>{headerHtml || ''}</div>
          </div>
        }
        <div className="payment-pinpad-body aaui-flexbox">
          <div className="payment-pinpad-body-img" title="iPP320 Device" />
          <div className="payment-pinpad-body-right">
            {hasError ?
              <div className="payment-pinpad-error-wrapper aaui-flexbox">
                <Alert type="error" className="payment-pinpad-error-alert" noClose>
                  <div className="error-content">
                    <div className="payment-pinpad-error-message">
                      <div>{this.getErrorMessage(errorMessage, actionType)}</div>
                    </div>
                    { (actionType !== actionTypes.HIDE_RETRY) &&
                      <div className="payment-pinpad-error-action">
                        <span className="payment-pinpad-retry" onClick={() => this.retry(false, paymentTypeId)}>Swipe card</span>
                        {paymentTypeId !== debitCardId ?
                          <span>
                            <span className="payment-pinpad-retry-separation">or</span>
                            <span className="payment-pinpad-retry" onClick={() => this.retry(true, paymentTypeId)}>Manual entry</span>
                          </span> : ''
                        }
                      </div>
                    }
                  </div>
                </Alert>
                {children || ''}
              </div> :
              <p className="payment-pinpad-body-prompts">{promptMessage}</p>
            }
          </div>
        </div>
      </div>
    );
  }

  renderWebServiceHelp = () => {
    const { webStartServicehelpURL } = this.props.pinpad.payment.toJS();

    return <WSIHelper url={webStartServicehelpURL} />;
  }

  getErrorMessage = (errorMessage, actionType) => {
    if (actionType === actionTypes.SERVICE_HELP) {
      return this.renderWebServiceHelp();
    }

    return errorMessage;
  }

  retry(isManulEntry, paymentType) {
    const { pinpad } = this.props;
    const { apdInterfaceApplet } = pinpad.pinpadModal.toJS();
    const { debitCardId, errorMessage, isNewCard } = pinpad.payment.toJS();
    const isDebitCard = paymentType === debitCardId;

    this.props.updateErrorMessage({
      message: '',
      hasError: false,
      actionType: actionTypes.HIDE_RETRY
    });

    /* istanbul ignore else */
    if (errorMessage) {
      this.props.updateErrorMessage({
        message: '',
        isServerErr: true
      });
    }

    this.props.showFailActionBtns(false);
    this.props.updateModalTitle(isNewCard ? titles.NEW_CARD_TITLE : titles.TRANSACTION_TITLE);
    setTimeout(() => { // wait the fail action take work
      apdInterfaceApplet.APDInputStart(isManulEntry, isDebitCard);
    }, 100);
  }
}

export default connect(
  null,
  {
    updateErrorMessage,
    showFailActionBtns,
    updateModalTitle
  }
)(Pinpad);
