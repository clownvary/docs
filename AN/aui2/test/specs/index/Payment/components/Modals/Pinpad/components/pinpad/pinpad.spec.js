import React from 'react';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';
import { fromJS } from 'immutable';
import configureStore from 'redux-mock-store';
import middlewares from 'shared/api/middlewares';
import { Pinpad } from 'index/Payment/components/Modals/PinPad/components/Pinpad/pinpad';

describe('index/Payment/components/Modals/PinPad/components/Pinpad/pinpad', () => {

  let store;

  beforeEach(() => {
    store = configureStore(middlewares)({});
  });
  const actions = {
    updateErrorMessage: jest.fn(),
    showFailActionBtns: jest.fn(),
    updateModalTitle: jest.fn()
  }
  const setup = (store, props) => mount(<Pinpad store={store} {...props} {...actions}></Pinpad>);

  const payment = {
    webStartServicehelpURL: '',
    receiptHeaderId: 1234,
    agentId: 0,
    isRefund: false,
    currentPayIndex: 0,
    promptMessage: "",
    isNewCard: false,
    cardHolderInfo: {
      cardholderZip: null,
      cardholderAddr: null
    },
    pays: [
      {
        ams_account_id: "APD AccountID",
        payment_type_id: 4,
        index: 0,
        payName: "Debit Card",
        amount: "370.00",
        apd_input_type: 1,
        paymentTypeId: 4
      }
    ],
    communicating: false,
    debitCardId: 4,
    companyId: 0,
    customerId: 1,
    apdAppletInfo: {
      device_type_id: 1,
      terminal_location_address: "937 Enterprise Dr",
      device_id: 999888,
      terminal_location_city: "Sacramento",
      terminal_location_state: "CA",
      terminal_location_zip: "95825",
      card_acceptor_name: "YMCA OF METROPOLITAN LOS ANGELES",
      merchant_descriptor: "ActiveNetTrainer",
      merchant_user: "AMS34VbVMSCRealDivId",
      merchant_password: "AMS34VbVMSCRealDivId",
      ams_url: "mstest.active.com",
      ams_port: 443,
      ams_timeout: 100,
      terminal_number: "007",
      usb_port_name: "COM5",
      pinpad_data_entry_timeout: 20,
      pinpad_swipe_card_timeout: 50,
      pinpad_pin_entry_timeout: 40,
      pinpad_device_timeout: 300,
      avs_full_addr: true,
      avs_zip: false,
      write_to_log: "Yes",
      applet_name: "APDInterfaceApplet",
      applets: "codebase=/device-libs/"
    },
    errorMessage: "pinpad error occur",
    apdPaymentInfo: {
      isPayerCashCustomer: false,
      hasImmediatePaymentDue: false,
      canCancelApdPayment: true,
      key: 1,
      committedReceiptId: 22916,
      receiptPaymentId: 20423,
      rno: 1,
      receiptNumber: 1000778.005,
      orderId: "Receipt ID 22916",
      orderDescriptor: "YMCA OF METROPOLITAN LOS ANGELES"
    }
  };
  const props = {
    pay: {
      ams_account_id: "APD AccountID",
      payment_type_id: 4,
      index: 0,
      payName: "Debit Card",
      amount: "370.00",
      apd_input_type: 1,
      paymentTypeId: 4
    },
    pinpad: {
      payment: fromJS(payment),
      pinpadModal: fromJS({
        title: "Processing Transaction",
        shown: true,
        apdInterfaceApplet: {
          APDInputStart: jest.fn()
        }
      }),
      pinpadFail: fromJS({
        shown: false
      })
    }
  };

  it('component should render correctly', () => {
    const component = setup(store, props);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('component should work fine', () => {
    jest.useFakeTimers();
    const pay = Object.assign({}, props.pay, {
      errorMessage: 'error by pinpad',
      hasError: true,
      actionType: 3,
      className: 'pinpad-class',
      headerHtml: 'xxxx in 2345'
    });
    const newProps = {
      ...props,
      pay,
      children: 'test pinpad'
    };
    const component = setup(store, newProps);
    const serviceHelp = component.find('.service-help');

    expect(component).toBeTruthy();
    expect(serviceHelp.length).toBe(1);
    serviceHelp.at(0).simulate('click');

    const pay2 = Object.assign({}, props.pay, {
      errorMessage: 'error by pinpad',
      hasError: true,
      paymentTypeId: 3
    });
    const newProps2 = {
      ...props,
      pay: pay2,
      children: ''
    };

    component.setProps(newProps2);
    component.find('.payment-pinpad-retry').at(0).simulate('click');
    jest.runAllTimers(100);
    jest.clearAllTimers();
    component.find('.payment-pinpad-retry').at(1).simulate('click');

    const pay3 = Object.assign({}, props.pay, {
      errorMessage: 'error by pinpad',
      hasError: true,
      paymentTypeId: 3
    })
    const newProps3 = {
      ...props,
      pay: pay3,
      pinpad: {
        ...props.pinpad,
        payment: fromJS({
          ...payment,
          isNewCard: true
        })
      },
      children: ''
    }
    component.setProps(newProps3);
    component.find('.payment-pinpad-retry').at(0).simulate('click');
    component.find('.payment-pinpad-retry').at(1).simulate('click');
    expect(component.find('.payment-pinpad-error-wrapper').length).toBe(1);
  });
});
