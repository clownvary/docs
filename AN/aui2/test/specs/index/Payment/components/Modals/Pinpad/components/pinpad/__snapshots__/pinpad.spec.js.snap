// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index/Payment/components/Modals/PinPad/components/Pinpad/pinpad component should render correctly 1`] = `
<Pinpad
  pay={
    Object {
      "amount": "370.00",
      "ams_account_id": "APD AccountID",
      "apd_input_type": 1,
      "index": 0,
      "payName": "Debit Card",
      "paymentTypeId": 4,
      "payment_type_id": 4,
    }
  }
  pinpad={
    Object {
      "payment": Immutable.Map {
        receiptHeaderId: 1234,
        agentId: 0,
        isRefund: false,
        currentPayIndex: 0,
        promptMessage: "",
        isNewCard: false,
        webStartServicehelpURL: "",
        cardHolderInfo: Immutable.Map {
            cardholderZip: null,
            cardholderAddr: null,
        },
        pays: Immutable.List [
            Immutable.Map {
                ams_account_id: "APD AccountID",
                payment_type_id: 4,
                index: 0,
                payName: "Debit Card",
                amount: "370.00",
                apd_input_type: 1,
                paymentTypeId: 4,
            },
        ],
        communicating: false,
        debitCardId: 4,
        companyId: 0,
        customerId: 1,
        apdAppletInfo: Immutable.Map {
            pinpad_data_entry_timeout: 20,
            avs_zip: false,
            avs_full_addr: true,
            write_to_log: "Yes",
            merchant_user: "AMS34VbVMSCRealDivId",
            device_id: 999888,
            merchant_descriptor: "ActiveNetTrainer",
            pinpad_pin_entry_timeout: 40,
            ams_timeout: 100,
            terminal_location_state: "CA",
            terminal_number: "007",
            terminal_location_address: "937 Enterprise Dr",
            usb_port_name: "COM5",
            card_acceptor_name: "YMCA OF METROPOLITAN LOS ANGELES",
            applet_name: "APDInterfaceApplet",
            terminal_location_city: "Sacramento",
            merchant_password: "AMS34VbVMSCRealDivId",
            ams_url: "mstest.active.com",
            device_type_id: 1,
            pinpad_swipe_card_timeout: 50,
            applets: "codebase=/device-libs/",
            pinpad_device_timeout: 300,
            ams_port: 443,
            terminal_location_zip: "95825",
        },
        errorMessage: "pinpad error occur",
        apdPaymentInfo: Immutable.Map {
            receiptNumber: 1000778.005,
            hasImmediatePaymentDue: false,
            orderId: "Receipt ID 22916",
            receiptPaymentId: 20423,
            rno: 1,
            committedReceiptId: 22916,
            canCancelApdPayment: true,
            isPayerCashCustomer: false,
            orderDescriptor: "YMCA OF METROPOLITAN LOS ANGELES",
            key: 1,
        },
    },
      "pinpadFail": Immutable.Map {
        shown: false,
    },
      "pinpadModal": Immutable.Map {
        title: "Processing Transaction",
        shown: true,
        apdInterfaceApplet: Immutable.Map {
            APDInputStart: [Function],
        },
    },
    }
  }
  showFailActionBtns={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "subscribe": [Function],
    }
  }
  updateErrorMessage={[Function]}
  updateModalTitle={[Function]}
>
  <div
    className="payment-pinpad "
    onClick={[Function]}
  >
    <div
      className="payment-pinpad-header"
    >
      <span>
        Debit Card
      </span>
      <span>
        $ 
        370.00
      </span>
      <div />
    </div>
    <div
      className="payment-pinpad-body aaui-flexbox"
    >
      <div
        className="payment-pinpad-body-img"
        title="iPP320 Device"
      />
      <div
        className="payment-pinpad-body-right"
      >
        <p
          className="payment-pinpad-body-prompts"
        />
      </div>
    </div>
  </div>
</Pinpad>
`;
